src/generic-container/started-generic-container.ts

```diff
@@ -11,20 +11,42 @@ import { Readable } from "stream";
 import { containerLogs } from "../docker/functions/container/container-logs";
 import { StoppedGenericContainer } from "./stopped-generic-container";
 import { stopContainer } from "../docker/functions/container/stop-container";
+import { restartContainer } from "../docker/functions/container/restart-container";
+import { WaitStrategy } from "../wait-strategy";
 
 export class StartedGenericContainer implements StartedTestContainer {
   constructor(
     private readonly container: Dockerode.Container,
     private readonly host: Host,
     private readonly inspectResult: InspectResult,
     private readonly boundPorts: BoundPorts,
-    private readonly name: ContainerName
+    private readonly name: ContainerName,
+    private readonly waitStrategy: WaitStrategy
   ) {}
 
   public async stop(options: Partial<StopOptions> = {}): Promise<StoppedTestContainer> {
     return this.stopContainer(options);
   }
 
+  public async restart(options: Partial<RestartOptions> = {}): Promise<StartedGenericContainer> {
+    const resolvedOptions: RestartOptions = { timeout: 0, ...options };
+    await restartContainer(this.container, resolvedOptions);
+    const newInspectRes = await inspectContainer(this.container);
+    const newBoundPorts = await BoundPorts.fromInspectResult(newInspectRes).filter(
+      Array.from(this.boundPorts.iterator()).map((port) => port[0])
+    );
+    await this.waitForContainer(this.container, newBoundPorts);
+
+    return new StartedGenericContainer(
```

Comment:
I wonder if the return type should be `Promise<void>` instead of `Promise<StartedGenericContainer>`. I say that because the new `StartedGenericContainer` is the same as the previous. Also, if I have a reference to the previous instance and for example `stop` it, the restarted container instance will also be stopped. I think therefore it should return `void` to make clear that you are working with the same container

Reply:
Hmm, I've decided to return a `StartedGenericContainer` because the bound ports was `readonly` and I kinda thought above conserving it to be that way, but now that I thought about it, it does make a lot more sense. Will make this change now.